### PR TITLE
feat: [DEL-3365]: Memoryff

### PR DESCRIPTION
**Pull request by Xin-Harness**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/31680)

### CODE REVIEW

**Review by arpit3107 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31680)* status: 14 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java, line 1844 at r1](https://reviewable.io/reviews/harness/harness-core/31680#-MyrczHdAnTz6dkQB-_V:-MyrczHdAnTz6dkQB-_W:blnutcj) ([raw file](https://github.com/harness/harness-core/blob/b4c2f59682ec87ba4b1679c72e08dfb5122554d0/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L1844)):*
> ```Java
>             .ciEnabled(false)
>             .delegateCpu(1)
>             .delegateRam(delegateRam)
> ```

earlier the delegateRam was 4 but now you without FF, ram will be 8
___
*[400-rest/src/test/resources/expectedDelegateWithoutDelegateName.sh, line 318 at r1](https://reviewable.io/reviews/harness/harness-core/31680#-Myre4PBECTInxiPPjjf:-Myre4PBECTInxiPPjjg:bbdl0o4) ([raw file](https://github.com/harness/harness-core/blob/b4c2f59682ec87ba4b1679c72e08dfb5122554d0/400-rest/src/test/resources/expectedDelegateWithoutDelegateName.sh#L318)):*
> ```Shell
> if [[ $DEPLOY_MODE == "KUBERNETES" ]]; then
>   echo "Starting delegate - version $2 with java $JRE_BINARY"
>   $JRE_BINARY $JAVA_OPTS $INSTRUMENTATION $PROXY_SYS_PROPS $OVERRIDE_TMP_PROPS -Ddelegatesourcedir="$DIR" -Xmx1536m -XX:+HeapDumpOnOutOfMemoryError -XX:+PrintGCDetails -XX:+PrintGCDateStamps -Xloggc:mygclogfilename.gc -XX:+UseParallelGC -XX:MaxGCPauseMillis=500 -Dfile.encoding=UTF-8 -Dcom.sun.jndi.ldap.object.disableEndpointIdentification=true -DLANG=en_US.UTF-8 -Xbootclasspath/p:alpn-boot-8.1.13.v20181017.jar -jar $2/delegate.jar config-delegate.yml watched $1
> ```

why this?



_Code quote:_
```Shell
-Xmx1536m
```

**Comment by raghuAtWings** [comment]
___
*[400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java, line 1844 at r1](https://reviewable.io/reviews/harness/harness-core/31680#-MyrczHdAnTz6dkQB-_V:-Mys-3UT5l9aYxVlzxPh:bp8ct0u) ([raw file](https://github.com/harness/harness-core/blob/b4c2f59682ec87ba4b1679c72e08dfb5122554d0/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L1844)):*
<details><summary><i>Previously, arpit3107 (Arpit Tiwari) wrote…</i></summary><blockquote>

earlier the delegateRam was 4 but now you without FF, ram will be 8
</blockquote></details>

thats correct, find for what kind of delegate we had ram to be 4

**Comment by raghuAtWings** [comment]
___
*[400-rest/src/test/resources/expectedDelegateOpenJdk.sh, line 337 at r2](https://reviewable.io/reviews/harness/harness-core/31680#-Mys-CT7DLXXYIsXI5tl:-Mys-CT83_rQGhzIv_L-:b33dksb) ([raw file](https://github.com/harness/harness-core/blob/c7012205b9c086c5952d1649f8c3bda0f9ec0833/400-rest/src/test/resources/expectedDelegateOpenJdk.sh#L337)):*
> ```Shell
> else
>   echo "Starting delegate - version $REMOTE_DELEGATE_VERSION with java $JRE_BINARY"
>   $JRE_BINARY $JAVA_OPTS $INSTRUMENTATION $PROXY_SYS_PROPS $OVERRIDE_TMP_PROPS -DACCOUNT_ID="ACCOUNT_ID" -DMANAGER_HOST_AND_PORT="https://localhost:9090" -Ddelegatesourcedir="$DIR" -Xmx4096m -XX:+HeapDumpOnOutOfMemoryError -XX:+PrintGCDetails -XX:+PrintGCDateStamps -Xloggc:mygclogfilename.gc -XX:+UseParallelGC -XX:MaxGCPauseMillis=500 -Dfile.encoding=UTF-8 -Dcom.sun.jndi.ldap.object.disableEndpointIdentification=true -DLANG=en_US.UTF-8 -Xbootclasspath/p:alpn-boot-8.1.13.v20181017.jar -jar delegate.jar config-delegate.yml watched $1
> ```

this should not change since by default the ff is off

**Comment by raghuAtWings** [comment]
___
*[400-rest/src/test/resources/expectedDelegateWithoutDelegateName.sh, line 318 at r1](https://reviewable.io/reviews/harness/harness-core/31680#-Myre4PBECTInxiPPjjf:-Mys-HgFAFhtcDnuAPc3:b-2ipum1) ([raw file](https://github.com/harness/harness-core/blob/b4c2f59682ec87ba4b1679c72e08dfb5122554d0/400-rest/src/test/resources/expectedDelegateWithoutDelegateName.sh#L318)):*
<details><summary><i>Previously, arpit3107 (Arpit Tiwari) wrote…</i></summary><blockquote>

why this?
</blockquote></details>

same here, this should not change since by default the ff is off

**Comment by raghuAtWings** [comment]
___
*[on-prem/kubernetes_installer/delegate.yaml, line 61 at r2](https://reviewable.io/reviews/harness/harness-core/31680#-Mys-sKUDg5yDv2Ffuql:-Mys-sKUDg5yDv2Ffuqm:b-qdxjeg) ([raw file](https://github.com/harness/harness-core/blob/0c3fac9d2e939e7f2b7822cbcb399134623833a4/on-prem/kubernetes_installer/delegate.yaml#L61)):*
> ```YAML
>             limits:
>               cpu: "1"
>               memory: "4Gi"
> ```

this should be also based on FF

**Review by raghuAtWings — CHANGES_REQUESTED** [review]
> Reviewed 12 of 14 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31680)* status: 2 change requests. 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[dockerization/delegate/immutable-scripts/start.sh, line 54 at r2](https://reviewable.io/reviews/harness/harness-core/31680#-Mys-YKv-hsodTTcpm5D:-Mys-YKv-hsodTTcpm5E:bcpdo7z) ([raw file](https://github.com/harness/harness-core/blob/0c3fac9d2e939e7f2b7822cbcb399134623833a4/dockerization/delegate/immutable-scripts/start.sh#L54)):*
> ```Shell
> 
> # 3. Start the delegate
> exec java $JAVA_OPTS $PROXY_SYS_PROPS -Xbootclasspath/p:alpn-boot-8.1.13.v20181017.jar -Xmx1536m -XX:+HeapDumpOnOutOfMemoryError -XX:+PrintGCDetails -XX:+PrintGCDateStamps -Xloggc:mygclogfilename.gc -XX:+UseParallelGC -XX:MaxGCPauseMillis=500 -Dfile.encoding=UTF-8 -Dcom.sun.jndi.ldap.object.disableEndpointIdentification=true -DLANG=en_US.UTF-8 -jar delegate.jar server config.yml
> ```

why this change?

immutab

**Review by Xin-Harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31680)* status: 2 change requests. 4. unresolved discussions. 0 of 1 approvals obtained. 

___
*[400-rest/src/test/resources/expectedDelegateOpenJdk.sh, line 337 at r2](https://reviewable.io/reviews/harness/harness-core/31680#-Mys-CT7DLXXYIsXI5tl:-Mys5cYe9cg7AOV8kgXH:b-ps6s4p) ([raw file](https://github.com/harness/harness-core/blob/c7012205b9c086c5952d1649f8c3bda0f9ec0833/400-rest/src/test/resources/expectedDelegateOpenJdk.sh#L337)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

this should not change since by default the ff is off
</blockquote></details>

We discussed this yesterday. Eventually the memory size will drop to 1536 so the expected size should be changed. I made the change to the test case to enable the feature flag so that the test case would pass.
___
*[dockerization/delegate/immutable-scripts/start.sh, line 54 at r2](https://reviewable.io/reviews/harness/harness-core/31680#-Mys-YKv-hsodTTcpm5D:-Mys6JB2BjzLXnM-C4nW:bo5bzah) ([raw file](https://github.com/harness/harness-core/blob/0c3fac9d2e939e7f2b7822cbcb399134623833a4/dockerization/delegate/immutable-scripts/start.sh#L54)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

why this change?

immutab
</blockquote></details>

Since the immutable delegate is designed to have smaller xmx so I changed.

**Review by Xin-Harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31680)* status: 8 file(s) to review. 2 change requests. 9. unresolved discussions. 0 of 1 approvals obtained. 

___
*[400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java, line 1844 at r1](https://reviewable.io/reviews/harness/harness-core/31680#-MyrczHdAnTz6dkQB-_V:-MysKk252MHKXJY28e_X:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/b4c2f59682ec87ba4b1679c72e08dfb5122554d0/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L1844)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

thats correct, find for what kind of delegate we had ram to be 4
</blockquote></details>

Done.
___
*[400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java, line 1767 at r2](https://reviewable.io/reviews/harness/harness-core/31680#-MyrzoHh91JQ5n-hPZhs:-MysKeSG-zRybAFccwNs:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0c3fac9d2e939e7f2b7822cbcb399134623833a4/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L1767)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

write this statement like below for readability

int delegateRam = featureFlagService.isEnabled(REDUCE_DELEGATE_MEMORY_SIZE, accountId) ? 4:8;
</blockquote></details>

Done.
___
*[400-rest/src/test/resources/expectedDelegateWithoutDelegateName.sh, line 318 at r1](https://reviewable.io/reviews/harness/harness-core/31680#-Myre4PBECTInxiPPjjf:-MysLA543bAWRetrZd-X:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/b4c2f59682ec87ba4b1679c72e08dfb5122554d0/400-rest/src/test/resources/expectedDelegateWithoutDelegateName.sh#L318)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

same here, this should not change since by default the ff is off
</blockquote></details>

Done.
___
*[400-rest/src/test/resources/expectedHarnessDelegate.yaml, line 68 at r2](https://reviewable.io/reviews/harness/harness-core/31680#-Mys-ORP2E3ire4Urwko:-MysLB0I9XTTijAb2iLF:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0c3fac9d2e939e7f2b7822cbcb399134623833a4/400-rest/src/test/resources/expectedHarnessDelegate.yaml#L68)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

same here
</blockquote></details>

Done.
___
*[400-rest/src/test/resources/expectedHarnessDelegateImmutable.yaml, line 68 at r2](https://reviewable.io/reviews/harness/harness-core/31680#-Mys-Q-x7X5sG-flQy-H:-MysLBc4D1k5gRx7yr8i:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0c3fac9d2e939e7f2b7822cbcb399134623833a4/400-rest/src/test/resources/expectedHarnessDelegateImmutable.yaml#L68)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

same here
</blockquote></details>

Done.
___
*[400-rest/src/test/resources/expectedHarnessDelegateWithCiEnabled.yaml, line 70 at r2](https://reviewable.io/reviews/harness/harness-core/31680#-Mys-Ryx5ab4jGjExcEH:-MysLCCdAK455v-L-shu:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0c3fac9d2e939e7f2b7822cbcb399134623833a4/400-rest/src/test/resources/expectedHarnessDelegateWithCiEnabled.yaml#L70)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

same here
</blockquote></details>

Done.
___
*[on-prem/kubernetes_installer/delegate.yaml, line 61 at r2](https://reviewable.io/reviews/harness/harness-core/31680#-Mys-sKUDg5yDv2Ffuql:-MysLCqm8KIecXOv95Bq:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0c3fac9d2e939e7f2b7822cbcb399134623833a4/on-prem/kubernetes_installer/delegate.yaml#L61)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

this should be also based on FF
</blockquote></details>

Done.

**Comment by Xin-Harness** [comment]
* install pr without the change and install delegates build 3565

test shell script
￼
￼
￼
￼

* upgrade the pr with the change and verify nothing changes build 3566
￼
￼

￼
￼
* 
* enable the FF to be true
￼
* create a new build with the same PR and upgrade and verify that xmx changes build 3567
￼
￼

￼
￼

**Comment by Xin-Harness** [comment]
Test:
* install pr without the change and install delegates build no.3565
test shell script
Install Delegate
image
kubernetes delegate
testk8s-kmpysm-0
Running
* upgrade the pr with the change and verify nothing changes build no.3566
Pasted Graphic 16
0 1509 pts0
Pasted Graphic 18

* enable the FF to be true
* 
Pasted Graphic 19
* create a new build with the same PR and upgrade and verify that xmx changes build no.3567
Pasted Graphic 20

Pasted Graphic 22
Pasted Graphic 23
Test successful

**Review by arpit3107 — DISMISSED** [review]
> Reviewed 6 of 14 files at r1, 8 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31680)* status: 1 change requests. 7. unresolved discussions.

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 7 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31680)* status: :shipit: complete! ✓

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 9 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31680)* status: :shipit: complete! ✓

File: 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java [hunk 1]
```diff
@@ -302,7 +302,8 @@
   AZURE_BLOB_SM,
   GIT_SYNC_PROJECT_CLEANUP,
   DISABLE_ARTIFACT_COLLECTION,
-  ENV_GROUP;
+  ENV_GROUP,
+  REDUCE_DELEGATE_MEMORY_SIZE;
```
Comment (gaurav-jain1) [inline 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java:306]: Consider renaming it to DELEGATE_MEMORY_SIZE_TO_1536M

**Pull request merged**